Add extra_data field to Package model #1532

JonoYang opened the pull request:
This PR reintroduces the `extra_data` field to the `Package` class.
It is being readded because we need a place to store the value for
`about_resource` so we can return it when `get_package_root()` is
called on a `Resource` that is an ABOUT file.

There are several other cases where keeping such extra data is
useful and needed for a specific Package type.

JonoYang committed:
Add extra_data to BasePackage again #1532

Signed-off-by: Jono Yang <jyang@nexb.com>

JonoYang committed:
Save about_resource to Package #2328

    * Get about_resource from Package in AboutPackage.get_package_root()

Signed-off-by: Jono Yang <jyang@nexb.com>

JonoYang committed:
Set AboutPackage.get_package_root to instance method

    * Instantiate packages in set_packages_root
    * Move extra_data field to Package

Signed-off-by: Jono Yang <jyang@nexb.com>

JonoYang committed:
Update test results #1532

Signed-off-by: Jono Yang <jyang@nexb.com>

JonoYang commented on src/packagedcode/about.py:
```diff
@@ -57,12 +57,8 @@ class AboutPackage(models.Package):
-        with io.open(manifest_resource.location, encoding='utf-8') as loc:
-            package_data = saneyaml.load(loc.read())
-        about_resource = package_data.get('about_resource')
+    def get_package_root(self, manifest_resource, codebase):
```
@pombredanne I've changed `get_package_root()` in `AboutPackage` from a classmethod to a regular instance method because I would like to access the `extra_data` field to get the `about_resource` value. I also modified https://github.com/nexB/scancode-toolkit/blob/31e3fee011dfb7fb671e9cdeb0ab7fc23f6c7a04/src/packagedcode/plugin_package.py#L137 to instantiate the package rather than just get the package class so the `extra_data` field is accessable. However, this introduces inconsistency on how the `get_package_root` method is called across packages, where this package requires the package instance to use the method whereas other packages do not.

pombredanne replied:
Hum... I had not noticed that part. Let me review first before merging then

pombredanne replied:
This has been fixed now ... I also improved get_package_instance significantly.

pombredanne committed:
Merge latest develop branch

Signed-off-by: Philippe Ombredanne <pombredanne@nexb.com>

pombredanne committed:
Add CHANGELOG entry for extra data #1532

Signed-off-by: Philippe Ombredanne <pombredanne@nexb.com>

pombredanne commented:
All green... lgtm and merging at last!

pombredanne committed:
Remove ignore_unknown arg from Package.create()

Also remove the exclude_properties arg.
Package.create() alywas only sets known attributes are set.

Signed-off-by: Philippe Ombredanne <pombredanne@nexb.com>

pombredanne committed:
Do not sort package keys when testing

This not needed anymore.

Signed-off-by: Philippe Ombredanne <pombredanne@nexb.com>

pombredanne committed:
Remove ignore_unknown arg from Package.create()

Also remove the exclude_properties arg.
Package.create() alywas only sets known attributes are set.
Format code and streamline docstrings

Signed-off-by: Philippe Ombredanne <pombredanne@nexb.com>

pombredanne committed:
De-serialize all objects with get_package_instance

Only a Package object was created before. We know also recreate nested
lists of dependencies/DependentPackage, parties/Party
and installed_files/File objects. This rehydrates a whole Package as
originally serialized.

Signed-off-by: Philippe Ombredanne <pombredanne@nexb.com>

pombredanne committed:
Keep Package.fields() method used in CSV output

Signed-off-by: Philippe Ombredanne <pombredanne@nexb.com>

pombredanne reviewed (approved):
All green. Merging now!
